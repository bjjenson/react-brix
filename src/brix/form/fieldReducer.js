export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validationResult: 'validationResult',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
}

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName, payload }) => {
  const handlers = {
    [actionTypes.updateValue]: value =>
      state.setIn([fieldName, current, 'value'], value)
        .setIn([fieldName, current, 'pristine'], value == state.getIn([fieldName, 'initial', 'value'])),

    [actionTypes.touched]: () =>
      state.setIn([fieldName, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) =>
      state.setIn([fieldName, current, 'error'], error)
        .setIn([fieldName, current, 'helperText'], helperText),

    [actionTypes.insertField]: fieldState =>
      state.set(fieldName, fieldState),

    [actionTypes.removeField]: () =>
      state.delete(fieldName),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {
    return handler(payload)
  }
  return state
}
