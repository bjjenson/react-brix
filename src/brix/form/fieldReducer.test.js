import { fromJS } from 'immutable'
import { fieldReducer, actions } from './fieldReducer'

let state, fieldName
beforeEach(() => {
  fieldName = 'field-name'
  state = fromJS({
    [fieldName]: {
      initial: {
        value: '',
      },
      current: {
        value: '',
        pristine: true,
        touched: false,
        error: false,
        helperText: '',
      },
    },
    otherField: {
      initial: {
        value: '',
      },
      current: {
        value: '',
        pristine: true,
        touched: false,
        error: false,
        helperText: '',
      },
    },
  })
})

test('type not handled, return state', () => {
  expect(fieldReducer(state, {})).toEqual(state)
})

test('updateValue for field updates value on current', () => {
  const action = actions.updateValue(fieldName, 'new value')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('pristine is false if value not equal to initial', () => {
  const action = actions.updateValue(fieldName, 'new value')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('pristine is true if value equal to initial', () => {
  state = state.setIn([fieldName, 'current', 'pristine'], false)
  const action = actions.updateValue(fieldName, '')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('touched set on current field', () => {
  const action = actions.touched(fieldName)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('validationResult set on current field', () => {
  const action = actions.validationResult(fieldName, true, 'i am error')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('insertField', () => {
  const fieldState = fromJS({
    initial: {
      value: 'one',
    },
    current: {
      value: 'one',
    },
  })
  const action = actions.insertField('newField', fieldState)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('removeField', () => {
  const action = actions.removeField(fieldName)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})
