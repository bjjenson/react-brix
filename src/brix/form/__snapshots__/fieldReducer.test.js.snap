// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insertField 1`] = `
Immutable.Map {
  "field-name": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
  "newField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "one",
    },
    "current": Immutable.Map {
      "value": "one",
    },
  },
}
`;

exports[`pristine is false if value not equal to initial 1`] = `
Immutable.Map {
  "field-name": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "new value",
      "pristine": false,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
}
`;

exports[`pristine is true if value equal to initial 1`] = `
Immutable.Map {
  "field-name": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
}
`;

exports[`removeField 1`] = `
Immutable.Map {
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
}
`;

exports[`touched set on current field 1`] = `
Immutable.Map {
  "field-name": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": true,
      "error": false,
      "helperText": "",
    },
  },
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
}
`;

exports[`updateValue for field updates value on current 1`] = `
Immutable.Map {
  "field-name": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "new value",
      "pristine": false,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
}
`;

exports[`validationResult set on current field 1`] = `
Immutable.Map {
  "field-name": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": true,
      "helperText": "i am error",
    },
  },
  "otherField": Immutable.Map {
    "initial": Immutable.Map {
      "value": "",
    },
    "current": Immutable.Map {
      "value": "",
      "pristine": true,
      "touched": false,
      "error": false,
      "helperText": "",
    },
  },
}
`;
